### ISSUE TITLE
Performance issue with big diffs in deepEqual

### ISSUE THREAD

**Issue by chrpeter** [tapjs/tapjs#699]
Hi.

I have a question/issue with big diffs in deepEqual comparison for an object.

I have a large object where I have multiple differences between actual/expected value. When the diff length gets longer, the response time increases a lot and I have to run tap with `--no-timeout` flag resulting in the test running for several minutes.
I've tried do the same thing in Jest and I get the same diff in matter of seconds.

With a simple test with generated data:
```javascript
import tap from 'tap';

const found = {};
const foundTestObject = {
    value: 'test',
    id: 'testId',
};

const wanted = {};
const wantedTestObject = {
    value: 'wantedTestObject',
    id: 'wantedTestId',
};

for(let i = 0; i<1000; i++){
  found[`field_${i}`] = foundTestObject;
  wanted[`field_${i}`] = wantedTestObject;
}

tap.deepEqual(found, wanted);
```

Jest uses approximately 3.4 seconds, whilst node-tap uses 60 seconds.

**Comment by isaacs** [comment]
This is a good test case, thanks.  I'll dig into it.

It looks like the inefficiency here is that it's generating a full tcompare formatted output string for both objects, and then doing a string diff on these two giant strings.  It'd be much more efficient to build up the diff along the way (which was how tcompare originally worked), but that's more complicated to do, since the current approach lets me just take an off-the-shelf string diffing library, and provides much better grouping and splitting of diffs where only one or two lines diverge in a very large output.  (Otherwise it's easy to miss where the actual difference was.)

So, if we return to that original approach, and ditch `jsdiff`, then it'll require a bit of work to get it back to the same level of prettiness.

For example of what I'm talking about, the old approach (which was pretty naive) would result in stuff like this:

```
       "field_999": Object {
    -    "value": "wantedTestObject",
    +    "value": "test",
    -    "id": "wantedTestId",
    +    "id": "testId",
       },
```

instead of this, which is much more readable:

```
       "field_999": Object {
    -    "value": "wantedTestObject",
    -    "id": "wantedTestId",
    +    "value": "test",
    +    "id": "testId",
       },
```

**Comment by isaacs** [comment]
Yeah, wow, it turns out that jsdiff is pretty slow with super long strings.  And since I already have a diff (albeit a clumsy one) coming out of the object walk, it's super expensive to then rip it apart and throw both strings at jsdiff to re-construct a new prettier diff.

I spent a few hours updating diff-frag to be smarter and more efficient, so I can get the exact same results by just walking the lines with a single `while (match = regexp.exec(diff))` and a few switch statements and state tracker.  Brings the >1 minute operation down to less than 1 second in the case you provided.  Will be fixed in tap 15.0.1, may backport to v14 if seems like it's easy to do.

**Comment by isaacs** [comment]
Ah, here's why that's tricky though:

Let's say you have some very large object where the _only_ thing that differs is the class name.  `tcompare` will create a diff that's something like:

```
--- expected
+++ actual
-ExpectedClass {
-  long: "list",
-  of: "many",
-  various: "fields",
-  ...
-}
+ActualClass {
+  long: "list",
+  of: "many",
+  various: "fields",
+  ...
+}
```

because it's a formatted object diff, rather than a naive string diff.

With jsdiff, you end up with a much nicer looking:

```
--- expected
+++ actual
-ExpectedClass {
+ActualClass {
   long: "list",
   of: "many",
   ...
```

So, in the cleanup, I need to be smarter about matching `+/-` lines and turning them into `' '` lines, at least to a rough approximation.

**Comment by isaacs** [comment]
Long time coming, but the next version of tcompare (which will be included in tap v18 #821) will have much better performance and UX for cases like these.  See tapjs/tcompare#20 for details.

**Comment by isaacs** [comment]
Tap v18 is now in prerelease. (Will be pushed to latest once some ts-node issues are resolved, until then, it's a slow install from git, but otherwise fine.)

`npm i tap@pre` to get it. Do make sure to read the [changelog](https://node-tap.org/changelog/#18.0), as a lot of things have been updated.
